### ISSUE TITLE
Spam account registrations

### ISSUE THREAD

**Issue by davidnicholls125** [johnvanbreda/naturespot-d8#144]
We had had multiple spam accounts opened in the last week or so with lots of junk added to the forums. This is a very recent issue, has anything changed in terms of the captcha and/or could we beef up the security?

**Comment by johnvanbreda** [comment]
I've not changed anything, but spambots are forever evolving unfortunately. I've switched to a maths style challenge on the user registration form which is the same as on iRecord, so let's see if that helps.

**Comment by davidnicholls125** [comment]
We had another this morning unfortunately. Is it possible perhaps to make one or more of the entry fields meet a particular format?
![image](https://user-images.githubusercontent.com/80914584/219852076-addd0013-bb7a-42d6-b63c-884e6c37ebb0.png)

**Comment by johnvanbreda** [comment]
I'm sure it's possible with a little bit of custom coding, maybe a couple of hours work. But whatever we do, in my experience a small number of spam accounts will still get through so I'd not be too concerned if one or two get through.

Another thought is to add a word blocklist to the forum post validation.

**Comment by davidnicholls125** [comment]
The blacklisting sounds worth a go - can you point me to the relevant admin page please?

**Comment by johnvanbreda** [comment]
There's no built in feature for the word blocking, so I've asked Biren for his advice as he has some experience of this sort of thing. There is a module https://www.drupal.org/project/protected_forms which should do the job but I'll wait to see what he says first.

**Comment by davidnicholls125** [comment]
Thanks. Another incident this morning but it appears to be all the same person/bot. The postcode entry seems always to be the same: 10001.

**Comment by davidnicholls125** [comment]
Did Biren have any advice? We are still having to do manual approvals which is a pain!

**Comment by johnvanbreda** [comment]
Yes, Biren's suggestion was to use this module: https://www.drupal.org/project/field_validation.

As it's installed via Composer, I should use the opportunity to upgrade all the Drupal modules to the latest 9.x code - probably best to let me know a good time as I'll want to take the site offline for a couple of hours.

**Comment by davidnicholls125** [comment]
Thanks John. I'm back now. I don't think it matters when the site is offline, though probably better during the day rather than the evening. It would be helpful to know when however so I can warn the trustees.

**Comment by johnvanbreda** [comment]
@davidnicholls125 I've spent today taking a copy of all the code and upgrading it to the latest of everything (sticking to Drupal 9.x as we haven't yet fully tested 10.x). As the upgrade is quite major I'd rather not touch the live site on a Friday, so the plan is on Monday morning I'll take the live site offline, take a full backup, replace the Drupal core, vendor and module folders and run the database upgrade scripts. I'll also update to PHP 8.1.

**Comment by davidnicholls125** [comment]
Thanks. I'll warn everyone!

**Comment by johnvanbreda** [comment]
Working on the upgrade now so expect the site to be offline at some point this morning.

**Comment by johnvanbreda** [comment]
The upgrade is now done. I've added the field validation module and set up simple validation rules for blocked words on a forum post body and the comment replies at:
* https://www.naturespot.org.uk/admin/structure/field_validation/manage/node_forum
* https://www.naturespot.org.uk/admin/structure/field_validation/manage/comment_comment_forum
 
Please update both these replies with your list of blocked words and also check the error message.

**Comment by davidnicholls125** [comment]
Not sure what specific words we would want to block but I'll not the links if anything comes to mind.
The main need was to prevent spam registrations in the first place. Currently we are having to approve each application manually, has anything changed in the security set up so that we can return to automatic approvals?

**Comment by johnvanbreda** [comment]
The main change, apart from upgrading modules to latest versions (which may or may-not improve anti-spam measures in itself) is that we now have the field_validation module. This allows us to set up a custom verification rule for any Drupal field - further up this thread we started discussing blocking registration attempts where field values were of a certain format, but the discussion moved onto word blocking which is what we have set up so far. Fortunately the field validation module does also allow us to take fine control over format validation. Looking at your example above, the quick and easy fix is for me to add a regular expression format validation so that postcodes have to be of the correct format (basically a pattern match). I'll do that now...
